File: solutions/javascript/complex-numbers/1/complex-numbers.js
/**
 * Class representing a complex number.
 */
export class ComplexNumber {
  /**
   * Create a complex number.
   * @param {number} real - The real part of the complex number.
   * @param {number} imag - The imaginary part of the complex number.
   * @throws Will throw an error if real or imag is not a number.
   */
  constructor(real, imag) {
    if (typeof real !== 'number' || typeof imag !== 'number') {
      throw new TypeError('Real and imaginary parts must be numbers');
    }
    this.real = real;
    this.imag = imag;
  }

  /**
   * Add a complex number to this complex number.
   * @param {ComplexNumber} other - The complex number to add.
   * @returns {ComplexNumber} The result of the addition.
   */
  add(other) {
    return new ComplexNumber(this.real + other.real, this.imag + other.imag);
  }

  /**
   * Subtract a complex number from this complex number.
   * @param {ComplexNumber} other - The complex number to subtract.
   * @returns {ComplexNumber} The result of the subtraction.
   */
  sub(other) {
    return new ComplexNumber(this.real - other.real, this.imag - other.imag);
  }

  /**
   * Multiply this complex number by another complex number.
   * @param {ComplexNumber} other - The complex number to multiply by.
   * @returns {ComplexNumber} The result of the multiplication.
   */
  mul(other) {
    return new ComplexNumber(
      this.real * other.real - this.imag * other.imag,
      this.imag * other.real + this.real * other.imag
    );
  }

  /**
   * Divide this complex number by another complex number.
   * @param {ComplexNumber} other - The complex number to divide by.
   * @returns {ComplexNumber} The result of the division.
   * @throws Will throw an error if division by zero occurs.
   */
  div(other) {
    const denominator = other.real * other.real + other.imag * other.imag;
    if (denominator === 0) {
      throw new Error('Division by zero');
    }
    return new ComplexNumber(
      (this.real * other.real + this.imag * other.imag) / denominator,
      (this.imag * other.real - this.real * other.imag) / denominator
    );
  }

  /**
   * Get the magnitude (absolute value) of this complex number.
   * @returns {number} The magnitude of the complex number.
   */
  get abs() {
    return Math.sqrt(this.real * this.real + this.imag * this.imag);
  }

  /**
   * Get the conjugate of this complex number.
   * @returns {ComplexNumber} The conjugate of the complex number.
   */
  get conj() {
    return new ComplexNumber(this.real, this.imag !== 0 ? -this.imag : 0);
  }

  /**
   * Get the exponential of this complex number.
   * @returns {ComplexNumber} The exponential of the complex number.
   */
  get exp() {
    return new ComplexNumber(
      Math.exp(this.real) * Math.cos(this.imag),
      Math.exp(this.real) * Math.sin(this.imag)
    );
  }
}

File: solutions/go/complex-numbers/1/complex_numbers.go
package complexnumbers

import "math"

type Number struct {
	real      float64
	imaginary float64
}

func (number Number) Real() float64 {
	return number.real
}

func (number Number) Imaginary() float64 {
	return number.imaginary
}

func (number Number) Add(n2 Number) Number {
	return Number{
		real:      number.real + n2.real,
		imaginary: number.imaginary + n2.imaginary,
	}
}

func (number Number) Subtract(n2 Number) Number {
	return Number{
		real:      number.real - n2.real,
		imaginary: number.imaginary - n2.imaginary,
	}
}

func (number Number) Multiply(n2 Number) Number {
	return Number{
		real:      number.real*n2.real - number.imaginary*n2.imaginary,
		imaginary: number.imaginary*n2.real + number.real*n2.imaginary,
	}
}

func (number Number) Times(factor float64) Number {
	return Number{
		real:      factor * number.real,
		imaginary: factor * number.imaginary,
	}
}

func (number Number) Divide(n2 Number) Number {
	divisor := n2.real*n2.real + n2.imaginary*n2.imaginary
	return Number{
		real:      (number.real*n2.real + number.imaginary*n2.imaginary) / divisor,
		imaginary: (number.imaginary*n2.real - number.real*n2.imaginary) / divisor,
	}
}

func (number Number) Conjugate() Number {
	return Number{
		real:      number.real,
		imaginary: -number.imaginary,
	}
}

func (number Number) Abs() float64 {
	return math.Sqrt(number.real*number.real + number.imaginary*number.imaginary)
}

func (number Number) Exp() Number {
	factor := math.Exp(number.real)
	return Number{
		real:      factor * math.Cos(number.imaginary),
		imaginary: factor * math.Sin(number.imaginary),
	}
}
